common-content/en/module/tools/first-nodejs-program/index.md

```diff
@@ -0,0 +1,114 @@
++++
+title = "Writing a NodeJS program"
+headless = true
+time = 60
+facilitation = false
+emoji= "🛠️"
+hide_from_overview = true
+objectives = [
+    "Write a zero-dependencies NodeJS program",
+]
++++
+
+Below we have a small NodeJS program. It is a bit like `wc`. It counts words in a file. Specifically, it counts words which contain a hyphen (`-`) character.
+
+It accepts one command line argument - the path of the file to read and count.
+
+Its output to stdout is just the number of words which contain a hyphen.
+
+It uses the same language (JavaScript) as we've written before, but uses some different APIs.
+
+```js
+import process from "node:process";
+import { promises as fs } from "node:fs";
+
+const argv = process.argv.slice(2);
+if (argv.length != 1) {
+    console.error(`Expected exactly 1 argument (a path) to be passed but got ${argv.length}.`);
+    process.exit(1);
+}
+const path = argv[0];
+
+const content = await fs.readFile(path, "utf-8");
+const wordsContainingHyphens = content.split(" ").filter((word) => word.indexOf("-") > -1).length;
+console.log(wordsContainingHyphens);
+```
+
+Let's play computer with this program - line by line:
+
+```js
+import process from "node:process";
+```
+
+This is loading some code from somewhere that isn't this file.
+
+We've seen `import` before. Here, instead of importing from a file we've written, we're importing the `process` module which is built into NodeJS.
+
+This is an example of the same language features (`import`) being used slightly differently (the `"node:"` is a special prefix to say "specially from node").
+
+The `process` module is built into NodeJS for managing our process. It can be used to do things like find out what arguments were passed to the process when it started, find out what user ran the process, exit the process, and more.
+
+```js
+import { promises as fs } from "node:fs";
+```
+
+We're importing another module.
+
+The `fs` module is built into NodeJS for interacting with the filesystem.
+
+This time, we're not importing the whole module. We are destructuring. The `node:fs` module exposes an object, and we are saying "import the `promises` property from the `fs` module, and bind it to the name `fs`".
+
+It's the equivalent to us writing `import { promises } from "node:fs"; const fs = promises;`.
+
+We are doing this because many of the things in the `fs` module don't support `async`/`await`, but `fs` has a sub-module called `promises` where everything supports `async`/`await`. Because we want to use `async`/`await`, we will use that. But having to write `fs.promises.readFile` is a bit annoying, so instead we import `fs.promises` as if it was just named `fs`.
```

Comment:
Even though I know what this means, I don't feel I quite know what this means. Shall we go round on it again?